crates/noirc_printable_type/src/lib.rs

```diff
@@ -0,0 +1,199 @@
+use std::{collections::BTreeMap, str};
+
+use acvm::FieldElement;
+use iter_extended::vecmap;
+use serde::{Deserialize, Serialize};
+
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(tag = "kind", rename_all = "lowercase")]
+pub enum PrintableType {
+    Field,
+    Array {
+        length: u64,
+        #[serde(rename = "type")]
+        typ: Box<PrintableType>,
+    },
+    SignedInteger {
+        width: u32,
+    },
+    UnsignedInteger {
+        width: u32,
+    },
+    Boolean,
+    Struct {
+        name: String,
+        fields: Vec<(String, PrintableType)>,
+    },
+    String {
+        length: u64,
+    },
+}
+
+impl PrintableType {
+    /// Returns the number of field elements required to represent the type once encoded.
+    pub fn field_count(&self) -> u32 {
+        match self {
+            Self::Field
+            | Self::SignedInteger { .. }
+            | Self::UnsignedInteger { .. }
+            | Self::Boolean => 1,
+            Self::Array { length, typ } => typ.field_count() * (*length as u32),
+            Self::Struct { fields, .. } => {
+                fields.iter().fold(0, |acc, (_, field_type)| acc + field_type.field_count())
+            }
+            Self::String { length } => *length as u32,
+        }
+    }
+}
+
+/// This is what all formats eventually transform into
+/// For example, a toml file will parse into TomlTypes
+/// and those TomlTypes will be mapped to Value
+#[derive(Debug, Clone, Serialize, PartialEq)]
+pub enum PrintableValue {
+    Field(FieldElement),
+    String(String),
+    Vec(Vec<PrintableValue>),
+    Struct(BTreeMap<String, PrintableValue>),
+}
+
+/// In order to display an `PrintableValue` we need an `PrintableType` to accurately
+/// convert the value into a human-readable format.
+pub struct PrintableValueDisplay<'a> {
+    value: &'a PrintableValue,
+    typ: &'a PrintableType,
+}
+
+impl<'a> PrintableValueDisplay<'a> {
+    #[must_use]
+    pub fn new(value: &'a PrintableValue, typ: &'a PrintableType) -> Self {
+        Self { value, typ }
+    }
+}
+
+impl<'a> std::fmt::Display for PrintableValueDisplay<'a> {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match (&self.value, &self.typ) {
+            (
+                PrintableValue::Field(f),
+                PrintableType::Field
+                // TODO: We should print the sign for these and probably print normal integers instead of field strings
```

Comment:
Could we make an issue for this TODO? Could be a good first issue for a new contributor

Reply:
```suggestion
                // TODO(#2401): We should print the sign for these and probably print normal integers instead of field strings
```